python/src/iceberg/conversions.py

```diff
@@ -0,0 +1,267 @@
+def _(primitive_type, value: int) -> bytes:
+    return struct.pack("<i", value)
+
+
+@to_bytes.register(LongType)
+@to_bytes.register(TimeType)
+@to_bytes.register(TimestampType)
+@to_bytes.register(TimestamptzType)
+def _(primitive_type, value: int) -> bytes:
+    return struct.pack("<q", value)
+
+
+@to_bytes.register(FloatType)
+def _(primitive_type, value: float) -> bytes:
+    return struct.pack("<f", value)
+
+
+@to_bytes.register(DoubleType)
+def _(primitive_type, value: float) -> bytes:
+    return struct.pack("<d", value)
+
+
+@to_bytes.register(StringType)
+def _(primitive_type, value: str) -> bytes:
+    return value.encode("UTF-8")
+
+
+@to_bytes.register(UUIDType)
+def _(primitive_type, value: uuid.UUID) -> bytes:
+    return struct.pack(">QQ", (value.int >> 64) & 0xFFFFFFFFFFFFFFFF, value.int & 0xFFFFFFFFFFFFFFFF)
+
+
+@to_bytes.register(BinaryType)
+@to_bytes.register(FixedType)
+def _(primitive_type, value: bytes) -> bytes:
+    return value
+
+
+@to_bytes.register(DecimalType)
+def _(primitive_type, value: Decimal) -> bytes:
+    value_as_tuple = value.as_tuple()
+    value_precision = len(value_as_tuple.digits)
+    value_scale = -value_as_tuple.exponent
+
+    if value_scale != primitive_type.scale:
+        raise ValueError(f"Cannot serialize value, scale of value does not match type {primitive_type}: {value_scale}")
+    elif value_precision > primitive_type.precision:
+        raise ValueError(
+            f"Cannot serialize value, precision of value is greater than precision of type {primitive_type}: {value_precision}"
+        )
+
+    with localcontext() as ctx:
+        ctx.prec = primitive_type.precision
+        unscaled_value = int((value * 10**primitive_type.scale).to_integral_value(rounding=ROUND_HALF_DOWN))
+        min_num_bytes = ((unscaled_value).bit_length() + 7) // 8
+        return unscaled_value.to_bytes(min_num_bytes, "big", signed=True)
+
+
+@singledispatch
+def from_bytes(primitive_type, b: bytes):
+    """A generic function which converts bytes to a built-in python value
+
+    Args:
+        primitive_type(PrimitiveType): An implementation of the PrimitiveType base class
+        b(bytes): The bytes to convert
+    """
+    raise TypeError(f"Cannot deserialize bytes, type {primitive_type} not supported: {str(b)}")
+
+
+@from_bytes.register(BooleanType)
+def _(primitive_type, b: bytes):
+    return struct.unpack("<?", b)[0] != 0
+
+
+@from_bytes.register(IntegerType)
+@from_bytes.register(DateType)
+def _(primitive_type, b: bytes):
+    return struct.unpack("<i", b)[0]
+
+
+@from_bytes.register(LongType)
+@from_bytes.register(TimeType)
+@from_bytes.register(TimestampType)
+@from_bytes.register(TimestamptzType)
+def _(primitive_type, b: bytes):
+    return struct.unpack("<q", b)[0]
+
+
+@from_bytes.register(FloatType)
+def _(primitive_type, b: bytes):
+    """
+    Note: float in python is implemented using a double in C. Therefore this involves a conversion of a 32-bit (single precision)
```

Comment:
I think this comment belongs on to_bytes, since the actual precision loss is going from double->float.  Float can be converted without loss to double.

Reply:
Ah right, that makes sense. Ok I'll move this there!

Reply:
Thinking about this some more, should this be described in both `to_bytes` and `from_bytes`? My understanding is that although the precision is lost when going `python float(64-bit) -> c float(32-bit)`, there's some imprecision that happens when going `c float(32-bit) -> python float(64-bit)`.
```c
int main(void) {
  int i;
  union {
    float f;
    unsigned char bytes[4];
  } packed;

  packed.f = 1.234;
  for (i=0; i<4; i++)
    printf ("\\x%02x", packed.bytes[i]);
  printf("\n");
}
```
output:
```
\xb6\xf3\x9d\x3f
```
```py
import struct

print(struct.unpack("<f", b"\xb6\xf3\x9d\x3f")[0])
```
output:
```
1.2339999675750732
```

Reply:
I guess it depends on your definition of precision. This arises because inability to actually represent 1.234 in ieee floats, so the  approximation differs between types.  This [stack-overflow post](https://stackoverflow.com/questions/17504833/why-converting-from-float-to-double-changes-the-value) I think gives a decent explanation.

```
>>> f"{1.2339999675750732:f}"
'1.234000'
>>> float(f"{1.2339999675750732:f}")
1.234
```

The only actual differences that I'm aware of between float and double is the bit-width for mantissa and exponent, both of which are strictly greater for double so the conversion should always be lossless, so in this case it would be precise.  How humans interpret that lossless conversion is interesting, and probably relevant for how filters are handled.

Printing doubles has some [research behind it](https://cseweb.ucsd.edu/~lerner/papers/fp-printing-popl16.pdf).  This is also why when people want to track currency, they use Decimal and not floats.

Reply:
Great, thanks for the explanation. That makes sense!

Reply:
Moved the note to `to_bytes`.